phoenix-core-client/src/main/java/org/apache/phoenix/expression/function/SubBinaryFunction.java

```diff
@@ -0,0 +1,214 @@
+            this.maxLength = maxLength >=0 ? maxLength : 0;
+        } else if (isOffsetConstant) {
+            Number offsetNumber = (Number)((LiteralExpression)getOffsetExpression()).getValue();
+            if (offsetNumber != null) {
+                int offset = offsetNumber.intValue();
+                PDataType type = getBinaryExpression().getDataType();
+                if (type.isFixedWidth()) {
+                    if (offset >= 0) {
+                        Integer maxLength = getBinaryExpression().getMaxLength();
+                        this.maxLength = maxLength - offset + (offset == 0 ? 0 : 1);
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public boolean evaluate(Tuple tuple, ImmutableBytesWritable ptr) {
+        Expression offsetExpression = getOffsetExpression();
+        if (!offsetExpression.evaluate(tuple, ptr)) {
+            return false;
+        }
+        if (ptr.getLength() == 0) {
+            return true;
+        }
+        int offset = offsetExpression.getDataType().getCodec().decodeInt(ptr, offsetExpression.getSortOrder());
+        int length = -1;
+        if (hasLengthExpression) {
+            Expression lengthExpression = getLengthExpression();
+            if (!lengthExpression.evaluate(tuple, ptr)) {
+                return false;
+            }
+            if (ptr.getLength() == 0) {
+                return true;
+            }
+            length = lengthExpression.getDataType().getCodec().decodeInt(ptr, lengthExpression.getSortOrder());
+            if (length <= 0) {
+                return false;
+            }
+        }
+        if (!getBinaryExpression().evaluate(tuple, ptr)) {
+            return false;
+        }
+        if (ptr.getLength()==0) {
+            return true;
+        }
+        byte[] bytes = new byte[]{};
+        int binLength;
+        if (getDataType() == PVarbinaryEncoded.INSTANCE) {
+            // get decoded bytes
+            bytes = (byte[]) PVarbinaryEncoded.INSTANCE.toObject(ptr.get(), ptr.getOffset(),
+                            ptr.getLength());
+            binLength = bytes.length;
+        } else {
+            binLength = ptr.getLength();
+        }
+        // Account for 1 versus 0-based offset
+        offset = offset - (offset <= 0 ? 0 : 1);
+        if (offset < 0) { // Offset < 0 means get from end
+            offset = binLength + offset;
+        }
+        if (offset < 0 || offset >= binLength) {
+            return false;
+        }
+        int maxLength = binLength - offset;
+        length = length == -1 ? maxLength : Math.min(length,maxLength);
+        if (getDataType() == PVarbinaryEncoded.INSTANCE) {
+            // get the slice of decoded bytes based on offset and length
+            byte[] result = Arrays.copyOfRange(bytes, offset, offset + length);
+            // encode the slice and set
+            ptr.set(PVarbinaryEncoded.INSTANCE.toBytes(result));
```

Comment:
Does `ptr.set(PVarbinaryEncoded.INSTANCE.toBytes(bytes, offset, offset + length))` work here instead of array copy?

Reply:
Do you mean `ptr.set(PVarbinaryEncoded.INSTANCE.toBytes(bytes), offset, length))`. That would not work since the offset and length are calculated on the decoded bytes.